src/clj/services/course_service/course_service.clj

```diff
@@ -23,34 +23,37 @@
 
 (s/fdef create-course-impl
         :args (s/cat :self #(satisfies? PCourseService %)
-                     :course-name :course/name)
-        :ret (s/keys :req [:course/id
-                           :course/name]))
+                     :course (s/keys :req [:course/name]))
+        :ret :course/course)
 
 
 (defn- create-course-impl
-  [this course-name]
-  (db/add-course! (.db this) course-name))
+  [this course]
+  (db/add-course! (.db this) course))
 
 
 (s/fdef validate-course-impl
         :args (s/cat :self #(= PCourseService (type %))
-                     :course-name :course/name)
-        :ret (s/map-of view/create-course-error-keys
-                       string?))
+                     :course-form-data map?)
+        :ret (s/or :success (s/keys :req [:course/name])
+                   :error (s/and map?
+                                 #(contains? % :errors))))
 
 
 (defn- validate-course-impl
-  [this course-name]
+  [this course-form-data]
   (let [courses (db/get-all-courses (.db this))
+        course-name (-> course-form-data
+                        :name
+                        (str/trim))
         courses-with-same-name (filter (fn [course] (= course-name (course :course/name))) courses)]
-    (-> {}
+    (-> {:course/name course-name}
         (#(if (s/valid? :course/name course-name)
             %
-            (assoc % :course/course-error "Der ausgewählte Name war inkorrekt!")))
+            (update % :errors merge {:course/course-error "The course name must be a non-empty string"})))
         (#(if (empty? courses-with-same-name)
             %
-            (assoc % :course/course-already-existed "Der angegebene Fachname wird bereits verwendet!"))))))
+            (update % :errors merge {:course/course-already-existed "The given course name already exists"}))))))
```

Comment:
Why did you use update merge, instead of simple assoc?

Reply:
we are setting a key in a nested map here, so if we have a second error message we need a merge for the `:errors` map.
this could have been `update % :errors assoc <k> <v>` as well, but I like this better

Reply:
Oh I understand, makes sense.